### ISSUE TITLE
unused: detect unexported struct fields that are initialised but never used

### ISSUE THREAD

**Issue by ainar-g** [dominikh/go-tools#288]
Sometimes after a refactoring, an unexported field in a structure remains initialised in a constructor, but is never used anywhere in package's methods or functions. Consider this code:

```go
type T struct {
	a int
}

func NewT(a int) T {
	return T{
		a: a,
	}
}

func main() {
	fmt.Println(NewT(42))
}
```

I would expect that unused will tell me that the field `a` is unused, but it doesn't, unless I comment the `a: a` in the constructor.

**Comment by ainar-g** [comment]
@EugeneOne1 has found an interesting variation of this with exported internal structs.  I wasn't sure if I should file a new issue, because it seems very related to me, but I can do so if you find that necessary.

In short, we have a struct like this:

```go
// pkg1 is an internal package.
package pkg1

// …

type Config struct {
        // …
        Field string `yaml:"-"`
        // …
}
```

Which is never read and only set once in another internal package:

```go
// pkg2 is another internal package.
package pkg2

// …

func Init() {
        // …
        var c pkg1.Config
        // …
        c.Field = someValue
        // …
}
```

**Comment by dominikh** [comment]
This gets complicated by reflection, cgo, and syscalls, all of which may provide reasons for writing to fields without seeing any reads in Go, even across package boundaries. I will have to spend more time thinking about this.

**Comment by dominikh** [comment]
Unrelated to those complications, here's an interesting example from database/sql's tests:

```
type fakeConn struct {
	...
	// Every operation writes to line to enable the race detector
	// check for data races.
	line int64
}

func (c *fakeConn) touchMem() {
	c.line++
}
```

While fairly esoteric, this is valid code that is only interested in the runtime effects of writing to the field, and the field is not unused. And while the post-increment also constitutes a read, this code actually discovered a bug in `unused`, where we only treated it as a write, because we believed that an increment without any other reads is pointless, which is very similar to considering field writes without reads pointless. And I believe that instead of an increment, this could've been a simple assignment and still have the same runtime effects.

**Comment by ainar-g** [comment]
I'm not against moving this particular check into the aggressive category, if that makes it easier.

**Comment by dominikh** [comment]
What I'll probably end up doing is provide a standalone tool that offers knobs for `unused`'s behavior, e.g. how to treat field writes or how to deal with exported APIs. U1000 in `staticcheck` will continue using conservative defaults.

**Comment by ainar-g** [comment]
To me, that sounds a bit like a duplication of effort, especially considering that `unused` used (heh) to be a separate tool but was later merged into `staticcheck`, and that the kind of person that would use `staticcheck --aggressive` would most probably also want to use `unused --include-fields`.  But you know the codebase better.

**Comment by dominikh** [comment]
> To me, that sounds a bit like a duplication of effort

Not really. Both U1000 and the standalone tool would use the same underlying code for computing usedness graphs, just that that code will have configuration-dependent branches.

We _could_ make this part of `aggressive`, but we can't do the same for flagging exported, unused API, another oft requested feature, because that needs a fundamentally different mode of operation from all other checks. Strictly speaking, the fact we can run U1000 in `staticcheck` is already a hack that requires special handling; it's not a true `go/analysis.Analyzer`. In the end, it was a mistake to merge `unused` with the other tools.

**Comment by dominikh** [comment]
A quick sneak peek:

```
$ cat foo.go
package pkg

type T struct {
	A int
	b int
	c int
}

func Generate() T {
	return T{
		A: 1,
		b: 2,
		c: 3,
	}
}

func Consume(t T) int {
	return t.b
}

$ unused -field-writes-are-uses=true foo.go 

$ unused -field-writes-are-uses=false foo.go
/home/dominikh/prj/src/example.com/foo.go:6:2: field c is unused
```

**Comment by dominikh** [comment]
Semi-rhetorical question: are the fields in the following example used?

```
type T struct {
	a int
	b int
}

func foo(m map[T]string) {
	key := T{1, 2}
	println(m[key])
}

```

If no, then that's a false positive. If yes, then what about `fmt.Sprintf` instead of map indexing? And if yes again, why `fmt.Sprintf` but not `fmt.Println`?
